core/src/test/java/com/google/instrumentation/common/ProviderTest.java

```diff
@@ -15,47 +15,111 @@
+    public GoodClass() {}
+  }
+
+  static class PrivateConstructorClass {
+    private PrivateConstructorClass() {}
+  }
+
+  static class NoDefaultConstructorClass {
+    public NoDefaultConstructorClass(int arg) {}
+  }
+
+  private static class PrivateClass {}
+
+  static interface MyInterface {}
+
+  static class MyInterfaceImpl implements MyInterface {
+    public MyInterfaceImpl() {}
+  }
+
   @Test
   public void testGoodClass() throws Exception {
-    GetGen getGen0 =
-        Provider.newInstance("com.google.instrumentation.common.ProviderTest$GetGen", null);
-    assertThat(getGen0).isNotNull();
-    assertThat(getGen0.getGen()).isEqualTo(0);
-    for (int i = 1; i < 10; i++) {
-      GetGen getGenI =
-          Provider.newInstance("com.google.instrumentation.common.ProviderTest$GetGen", null);
-      assertThat(getGenI).isNotNull();
-      assertThat(getGenI.getGen()).isEqualTo(i);
-    }
-    assertThat(getGen0.getGen()).isEqualTo(0);
+    assertThat(
+            Provider.<GoodClass>newInstance(
+                "com.google.instrumentation.common.ProviderTest$GoodClass", null))
+        .isNotNull();
   }
 
   @Test
   public void testBadClass() throws Exception {
-    GetGen getGen =
-        Provider.newInstance("com.google.instrumentation.common.ProviderTest$BadClass", null);
-    assertThat(getGen).isNull();
+    assertThat(
+            Provider.<GoodClass>newInstance(
+                "com.google.instrumentation.common.ProviderTest$BadClass", null))
+        .isNull();
   }
 
-  static class GetGen {
-    static int genCount = 0;
-    int gen;
+  @Test(expected = ServiceConfigurationError.class)
+  public void createInstance_ThrowsErrorWhenPrivateClass() throws Exception {
+    Provider.createInstance(
+        Class.forName(
+            "com.google.instrumentation.common.ProviderTest$PrivateClass",
+            true,
+            Provider.getCorrectClassLoader(ProviderTest.class)),
+        PrivateClass.class);
+  }
+
+  @Test(expected = ServiceConfigurationError.class)
+  public void createInstance_ThrowsErrorWhenPrivateConstructor() throws Exception {
+    Provider.createInstance(
+        Class.forName(
+            "com.google.instrumentation.common.ProviderTest$PrivateConstructorClass",
+            true,
+            Provider.getCorrectClassLoader(ProviderTest.class)),
+        PrivateConstructorClass.class);
+  }
+
+  @Test(expected = ServiceConfigurationError.class)
+  public void createInstance_ThrowsErrorWhenNoDefaultConstructor() throws Exception {
+    Provider.createInstance(
+        Class.forName(
+            "com.google.instrumentation.common.ProviderTest$NoDefaultConstructorClass",
+            true,
+            Provider.getCorrectClassLoader(ProviderTest.class)),
+        NoDefaultConstructorClass.class);
+  }
 
-    public GetGen() {
-      gen = genCount++;
-    }
+  @Test(expected = ServiceConfigurationError.class)
+  public void createInstance_ThrowsErrorWhenNoSubclass() throws Exception {
```

Comment:
I think this test name could be clearer, for example, `createInstance_ThrowsErrorWhenClassIsNotASubclass` or `createInstance_ThrowsErrorWhenClassCannotBeCast`.